File: src/Nav/compressed_telemtry/compressed_telemtry/gridmap_decompressor.py
import rclpy
from rclpy.node import Node
from grid_map_msgs.msg import GridMap
from std_msgs.msg import UInt8MultiArray, Float32MultiArray
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy, HistoryPolicy
import numpy as np
import struct
import zlib


class GridMapDecompressor(Node):
    PROTOCOL_VERSION = 0x01

    def __init__(self):
        super().__init__("gridmap_decompressor")

        radio_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )

        rviz_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )

        self.pub = self.create_publisher(GridMap, "/viz/traversability_map", rviz_qos)

        self.sub = self.create_subscription(
            UInt8MultiArray,
            "/telemetry/gridmap_compressed",
            self.map_callback,
            radio_qos,
        )

    def map_callback(self, msg):
        try:
            # 1. Unzip bytes
            compressed_bytes = bytes(msg.data)
            raw_payload = zlib.decompress(compressed_bytes)

            # 2. Validate minimum header size before unpacking.
            struct_size = struct.calcsize("<2B 10d 2H")
            if len(raw_payload) < struct_size:
                self.get_logger().warning(
                    "Dropped gridmap packet: payload smaller than header."
                )
                return

            # 3. Extract metadata and verify protocol version.
            meta = struct.unpack("<2B 10d 2H", raw_payload[:struct_size])
            protocol_version = meta[0]
            if protocol_version != self.PROTOCOL_VERSION:
                self.get_logger().warning(
                    f"Dropped gridmap packet: protocol version mismatch "
                    f"({protocol_version} != {self.PROTOCOL_VERSION})."
                )
                return

            frame_len = meta[1]
            resolution, length_x, length_y = meta[2], meta[3], meta[4]
            px, py, pz, qx, qy, qz, qw = (
                meta[5],
                meta[6],
                meta[7],
                meta[8],
                meta[9],
                meta[10],
                meta[11],
            )
            outer_start, inner_start = meta[12], meta[13]

            # 4. Extract Frame ID string
            frame_start = struct_size
            frame_end = struct_size + frame_len
            if len(raw_payload) < frame_end:
                self.get_logger().warning(
                    "Dropped gridmap packet: truncated frame_id bytes."
                )
                return
            frame_id = raw_payload[frame_start:frame_end].decode("utf-8")

            # 5. Validate alignment and split remaining bytes.
            # Every cell has 2 bytes for int16 elevation, 1 byte for uint8 traversability = 3 bytes total
            remaining_bytes = raw_payload[frame_end:]
            if len(remaining_bytes) % 3 != 0:
                self.get_logger().warning(
                    "Dropped gridmap packet: payload alignment is invalid."
                )
                return
            num_cells = len(remaining_bytes) // 3

            elev_bytes = remaining_bytes[: num_cells * 2]
            trav_bytes = remaining_bytes[num_cells * 2 :]

            # 6. Load back into Numpy
            # Elevation samples are encoded on-wire as little-endian int16.
            elev_int16 = np.frombuffer(elev_bytes, dtype=np.dtype("<i2"))
            trav_uint8 = np.frombuffer(trav_bytes, dtype=np.uint8)

            # 7. Reconstruct Floats and restore NaNs
            elev_float = np.where(
                elev_int16 == -32768, np.nan, elev_int16 / 1000.0
            ).astype(np.float32)
            trav_float = np.where(
                trav_uint8 == 255, np.nan, trav_uint8.astype(np.float32) / 100.0
            )

            # 8. Build the fresh ROS 2 GridMap message
            grid_msg = GridMap()
            grid_msg.header.stamp = self.get_clock().now().to_msg()  # Fix desync
            grid_msg.header.frame_id = frame_id

            grid_msg.info.resolution = resolution
            grid_msg.info.length_x = length_x
            grid_msg.info.length_y = length_y
            grid_msg.info.pose.position.x = px
            grid_msg.info.pose.position.y = py
            grid_msg.info.pose.position.z = pz
            grid_msg.info.pose.orientation.x = qx
            grid_msg.info.pose.orientation.y = qy
            grid_msg.info.pose.orientation.z = qz
            grid_msg.info.pose.orientation.w = qw

            grid_msg.layers = ["elevation", "traversability"]
            grid_msg.basic_layers = []  # Usually empty for basic visualization
            grid_msg.outer_start_index = outer_start
            grid_msg.inner_start_index = inner_start

            # Package arrays into Float32MultiArray wrappers
            elev_multi = Float32MultiArray()
            elev_multi.data = elev_float.tolist()

            trav_multi = Float32MultiArray()
            trav_multi.data = trav_float.tolist()

            grid_msg.data = [elev_multi, trav_multi]

            # 9. Render in RViz!
            self.pub.publish(grid_msg)

        except Exception as e:
            self.get_logger().error(f"Decompression failed: {e}")


def main(args=None):
    rclpy.init(args=args)
    node = GridMapDecompressor()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/Nav/compressed_telemetry_cpp/src/GridmapDecompressor.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <functional>
#include <limits>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include <zlib.h>

#include "grid_map_msgs/msg/grid_map.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_components/register_node_macro.hpp"
#include "std_msgs/msg/float32_multi_array.hpp"
#include "std_msgs/msg/multi_array_dimension.hpp"
#include "std_msgs/msg/multi_array_layout.hpp"
#include "std_msgs/msg/u_int8_multi_array.hpp"

namespace compressed_telemetry_cpp {

namespace {
constexpr uint8_t kProtocolVersion = 0x01;
constexpr size_t kHeaderSize =
    2 + (10 * sizeof(double)) + (2 * sizeof(uint16_t));

std::vector<uint8_t> DecompressZlib(const uint8_t *data, size_t size) {
  constexpr size_t kMaxDecompressed = 100 * 1024 * 1024;
  size_t output_size = std::max<size_t>(size * 4, 1024);

  for (int attempt = 0; attempt < 8 && output_size <= kMaxDecompressed;
       ++attempt) {
    std::vector<uint8_t> output(output_size);
    uLongf dest_len = output_size;
    int ret = uncompress(reinterpret_cast<Bytef *>(output.data()), &dest_len,
                         reinterpret_cast<const Bytef *>(data), size);
    if (ret == Z_OK) {
      output.resize(dest_len);
      return output;
    }
    if (ret != Z_BUF_ERROR) {
      throw std::runtime_error("zlib decompression failed");
    }
    output_size *= 2;
  }

  throw std::runtime_error("zlib decompression exceeded size limits");
}

template <typename T>
T ReadLittleEndian(const std::vector<uint8_t> &buffer, size_t offset) {
  T value{};
  std::memcpy(&value, buffer.data() + offset, sizeof(T));
#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
  uint8_t *raw = reinterpret_cast<uint8_t *>(&value);
  std::reverse(raw, raw + sizeof(T));
#endif
  return value;
}
} // namespace

class GridmapDecompressor : public rclcpp::Node {
public:
  explicit GridmapDecompressor(
      const rclcpp::NodeOptions &options = rclcpp::NodeOptions())
      : rclcpp::Node("gridmap_decompressor", options) {
    auto radio_qos = rclcpp::QoS(1).best_effort().durability_volatile();
    auto rviz_qos = rclcpp::QoS(1).reliable().transient_local();

    pub_ = this->create_publisher<grid_map_msgs::msg::GridMap>(
        "/viz/traversability_map", rviz_qos);
    sub_ = this->create_subscription<std_msgs::msg::UInt8MultiArray>(
        "/telemetry/gridmap_compressed", radio_qos,
        std::bind(&GridmapDecompressor::mapCallback, this,
                  std::placeholders::_1));
  }

private:
  void mapCallback(const std_msgs::msg::UInt8MultiArray::SharedPtr msg) {
    try {
      auto raw_payload = DecompressZlib(msg->data.data(), msg->data.size());
      if (raw_payload.size() < kHeaderSize) {
        RCLCPP_WARN(this->get_logger(),
                    "Dropped gridmap packet: payload smaller than header.");
        return;
      }

      uint8_t protocol_version = raw_payload[0];
      if (protocol_version != kProtocolVersion) {
        RCLCPP_WARN(
            this->get_logger(),
            "Dropped gridmap packet: protocol version mismatch (%u != %u).",
            protocol_version, kProtocolVersion);
        return;
      }

      const uint8_t frame_len = raw_payload[1];
      size_t offset = 2;

      const double resolution = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double length_x = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double length_y = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);

      const double px = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double py = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double pz = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);

      const double qx = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double qy = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double qz = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);
      const double qw = ReadLittleEndian<double>(raw_payload, offset);
      offset += sizeof(double);

      const uint16_t outer_start =
          ReadLittleEndian<uint16_t>(raw_payload, offset);
      offset += sizeof(uint16_t);
      const uint16_t inner_start =
          ReadLittleEndian<uint16_t>(raw_payload, offset);
      offset += sizeof(uint16_t);

      if (raw_payload.size() < offset + frame_len) {
        RCLCPP_WARN(this->get_logger(),
                    "Dropped gridmap packet: truncated frame_id bytes.");
        return;
      }

      const std::string frame_id(
          reinterpret_cast<const char *>(raw_payload.data() + offset),
          frame_len);
      offset += frame_len;

      const size_t remaining_bytes = raw_payload.size() - offset;
      if (remaining_bytes % 3 != 0) {
        RCLCPP_WARN(this->get_logger(),
                    "Dropped gridmap packet: payload alignment is invalid.");
        return;
      }

      const size_t num_cells = remaining_bytes / 3;
      const size_t elev_bytes_len = num_cells * sizeof(int16_t);
      if (remaining_bytes < elev_bytes_len) {
        RCLCPP_WARN(this->get_logger(),
                    "Dropped gridmap packet: elevation bytes missing.");
        return;
      }

      const uint8_t *elev_bytes = raw_payload.data() + offset;
      const uint8_t *trav_bytes = raw_payload.data() + offset + elev_bytes_len;

      std::vector<float> elev_float(num_cells);
      std::vector<float> trav_float(num_cells);

      for (size_t i = 0; i < num_cells; ++i) {
        int16_t elev_val;
        std::memcpy(&elev_val, elev_bytes + i * sizeof(int16_t),
                    sizeof(int16_t));
#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
        elev_val = static_cast<int16_t>((static_cast<uint16_t>(elev_val) >> 8) |
                                        (static_cast<uint16_t>(elev_val) << 8));
#endif
        if (elev_val == -32768) {
          elev_float[i] = std::numeric_limits<float>::quiet_NaN();
        } else {
          elev_float[i] = static_cast<float>(elev_val) / 1000.0f;
        }

        const uint8_t trav_val = trav_bytes[i];
        if (trav_val == 255) {
          trav_float[i] = std::numeric_limits<float>::quiet_NaN();
        } else {
          trav_float[i] = static_cast<float>(trav_val) / 100.0f;
        }
      }

      grid_map_msgs::msg::GridMap grid_msg;
      const auto now = this->get_clock()->now();
      const auto now_ns = now.nanoseconds();
      grid_msg.header.stamp.sec = static_cast<int32_t>(now_ns / 1000000000LL);
      grid_msg.header.stamp.nanosec =
          static_cast<uint32_t>(now_ns % 1000000000LL);
      grid_msg.header.frame_id = frame_id;

      grid_msg.info.resolution = resolution;
      grid_msg.info.length_x = length_x;
      grid_msg.info.length_y = length_y;
      grid_msg.info.pose.position.x = px;
      grid_msg.info.pose.position.y = py;
      grid_msg.info.pose.position.z = pz;
      grid_msg.info.pose.orientation.x = qx;
      grid_msg.info.pose.orientation.y = qy;
      grid_msg.info.pose.orientation.z = qz;
      grid_msg.info.pose.orientation.w = qw;

      grid_msg.layers = {"elevation", "traversability"};
      grid_msg.basic_layers = {};
      grid_msg.outer_start_index = outer_start;
      grid_msg.inner_start_index = inner_start;

      int size_x =
          std::max(1, static_cast<int>(std::round(length_x / resolution)));
      int size_y =
          std::max(1, static_cast<int>(std::round(length_y / resolution)));

      if (static_cast<size_t>(size_x) * static_cast<size_t>(size_y) !=
          num_cells) {
        RCLCPP_WARN(this->get_logger(),
                    "Gridmap dimensions do not match payload size (%zu cells "
                    "vs %d x %d).",
                    num_cells, size_x, size_y);
      }

      std_msgs::msg::MultiArrayLayout layout;
      layout.dim.resize(2);
      layout.dim[0].label = "column_index";
      layout.dim[0].size = static_cast<uint32_t>(size_x);
      layout.dim[0].stride = static_cast<uint32_t>(size_y);
      layout.dim[1].label = "row_index";
      layout.dim[1].size = static_cast<uint32_t>(size_y);
      layout.dim[1].stride = 1;
      layout.data_offset = 0;

      std_msgs::msg::Float32MultiArray elev_multi;
      elev_multi.layout = layout;
      elev_multi.data = std::move(elev_float);

      std_msgs::msg::Float32MultiArray trav_multi;
      trav_multi.layout = layout;
      trav_multi.data = std::move(trav_float);

      grid_msg.data = {std::move(elev_multi), std::move(trav_multi)};

      pub_->publish(grid_msg);
    } catch (const std::exception &e) {
      RCLCPP_ERROR(this->get_logger(), "Decompression failed: %s", e.what());
    }
  }

  rclcpp::Publisher<grid_map_msgs::msg::GridMap>::SharedPtr pub_;
  rclcpp::Subscription<std_msgs::msg::UInt8MultiArray>::SharedPtr sub_;
};

} // namespace compressed_telemetry_cpp

RCLCPP_COMPONENTS_REGISTER_NODE(compressed_telemetry_cpp::GridmapDecompressor)
